neps/nep-0480.md

```diff
@@ -0,0 +1,137 @@
+
+- Permissions: It could be possible to set permissions on a particular namespace. For example, restricting the namespace `dao_multisig` to interactions with `dao.near` only.
+
+- Synchronous execution. Namespaced contracts on the same account should be able to communicate with each other synchronously instead of asynchronously.
+
+## Specification
+
+### The `Namespace` type
+
+We introduce a new primitive type: `Namespace`. A `Namespace` is a string that contains only lowercase letters, numbers, and underscores. The empty string is considered the default namespace.
+
+### `DeployContract` action
+
+The `DeployContract` action is modified to include a `namespace` field. If the `namespace` field is not specified in an RPC call, the default namespace (the empty string) is used. The attached code is then deployed to the specified namespace.
+
+If code has already been deployed to the specified namespace, it is replaced, leaving code deployed to other namespaces untouched.
+
+Once a namespace is created, this NEP does not provide a way to delete it. However, namespaces should not be considered permanent.
+
+### `FunctionCall` action
+
+The `FunctionCall` action is modified to include a `namespace` field. If the `namespace` field is not specified in an RPC call, the default namespace (the empty string) is used. The function call is then executed on the code deployed to the specified namespace.
+
+### `FunctionCallPermission`
+
+A `namespaces` field is added to `FunctionCallPermission`. This field contains a list of namespaces that the access key is allowed to call. If the `namespaces` field is not specified, the access key is allowed to call all namespaces.
+
+### NEAR VM host functions
+
+The following host functions are added to the NEAR host environment:
+
+- `predecessor_namespace() -> Namespace`: Returns the namespace of the predecessor account, which may be the empty namespace. In the case that the predecessor is not a smart contract, this is the empty namespace.
+- `current_namespace() -> Namespace`: Returns the namespace of the current account, which may be the empty namespace.
+- There is no need for a `signer_namespace` function, because a signer is never a smart contract, and can therefore never have a non-empty namespace.
+
+Some existing host functions will need to be updated to include namespaces:
+
+- `promise_batch_action_function_call`: Updated to include namespace.
+- `promise_batch_action_deploy_contract`: Updated to include namespace.
+- etc.
+
+Existing host functions will generally still refer to the account as a whole:
+
+- `storage_usage() -> u64`: Returns the storage usage of the current account, including all namespaces.
+
+### RPC view calls
+
+#### `view_account`
+
+The `code_hash` field contains the hash of the code deployed to the default namespace. The `code_hashes` field contains a map of namespaces to code hashes.
```

Comment:
Should this clarify if `code_hashes` map will also include the default namespace or not?